templates/project/res/xml/opener_paths.xml

```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- https://developer.android.com/reference/android/support/v4/content/FileProvider.html#SpecifyFiles -->
+<paths xmlns:android="http://schemas.android.com/apk/res/android">
```

Comment:
Should this file be managed as part of the cordova-file-plugin (which provides API to exposes those paths) or is this also exposed by cordova-android directly?

Reply:
hmm, so the file needs to exist, and probably needs at least one path defined...

I guess one option would be for cordova-android to provide a default (mostly empty) implementation of this file, and then we update the cordova-plugin-file to use `edit-config` to add the rest of the paths at install time. That would only really work with major bumps to both cordova-android and cordova-plugin-file, but we're looking at a major for this repo anyways.

@erisu @breautek Thoughts? Are there downsides or security risks to including all these paths here by default? I'm not too familiar with Android's FileProvider stuff 😕

Reply:
> I guess one option would be for cordova-android to provide a default (mostly empty) implementation of this file, and then we update the cordova-plugin-file to use edit-config to add the rest of the paths at install time.

As much as having cordova-plugin-file actually handle this kind of stuff, and it using `<edit-config>` to change the config where necessary might be the "correct" approach... `edit-config` directive is incredibly broken... in the sense that it's overly sensitive to detecting conflicts. So if two things are editing the same node, there's a high chance they won't work or be compatible together, even if they aren't touching or editing the same attribute.

As for security risks... I'm not sure, I'm not entirely proficient in this area of Android either. So I'll need to do some reading. I do believe setting up a `FileProvider` does open the door to sharing files to third-party applications, which might be undesirable for some applications, especially if it's allowing `dataDirectory` files to be shared (which is normally a private storage container). But I'm not stating this with any kind of certainty...

I think to **actually** share/expose a file to another app, the application must use [Context.grantUriPermission](https://developer.android.com/reference/android/content/Context.html#grantUriPermission(java.lang.String,%20android.net.Uri,%20int)) on the `content://` url. So just by having a FileProvider available might not be enough to share files, I think this needs to be confirmed.

Reply:
Who might know the security implications of having a FileProvider like this?

Reply:
On the security front, I think we are in the clear.

Having the file provider and declaring which files/folders are shareable/visible isn't enough to share files. It is part of the process to setup fire sharing, but the app needs to do additional steps to actually share files.

Most of the time this is done via setting up a specialized [Activity](https://developer.android.com/training/secure-file-sharing/share-file), where a third-party app would create an Intent, which may target your app in which your app will start an activity to select a file and provide a result. The `Context.grantUriPermission` can achieve the same result, but permisison will be granted permanently, or until the app explicitly revokes the permission, whereas the Intent way is a temporary permission grant.

Cordova apps only have a single activity, generally speaking, which is the main activity for the application. So I believe the only way this can be an issue is if someone decides to implement an activity that responds to intents and uses this file provider being introduced in this PR, which is an unlikely scenario.

We could write a note for future app developers/plugin authors in `templates/project/res/xml/opener_paths.xml` that describing that the file provider may only be used for the application itself and should **not** be used for responding to Intents because it will expose the app's private data folders.

Reply:
Great idea.  I added a comment to the `cdv_core_file_provider_paths.xml` file that reads:

    Note: This File provider may only be used by the Cordova core itself
    and should not be used for responding to Intents because it will
    expose the app's private data folders.

Does that sound OK?